src/controllers/containerapp.go

```diff
@@ -0,0 +1,230 @@
+			Namespace:   capp.Metadata.Namespace,
+			Annotations: convertAnnotationsToMap(capp.Annotations),
+			Labels:      convertLabelsToMap(capp.Labels),
+		},
+		Spec: v1alpha1.CappSpec{
+			ScaleMetric:       capp.Spec.ScaleMetric,
+			Site:              capp.Spec.Site,
+			State:             capp.Spec.State,
+			ConfigurationSpec: capp.Spec.ConfigurationSpec,
+			RouteSpec:         capp.Spec.RouteSpec,
+			LogSpec:           capp.Spec.LogSpec,
+			VolumesSpec:       capp.Spec.VolumesSpec,
+		},
+		Status: v1alpha1.CappStatus{
+			ApplicationLinks:    capp.Status.ApplicationLinks,
+			KnativeObjectStatus: capp.Status.KnativeObjectStatus,
+			RevisionInfo:        capp.Status.RevisionInfo,
+			StateStatus:         capp.Status.StateStatus,
+			LoggingStatus:       capp.Status.LoggingStatus,
+			RouteStatus:         capp.Status.RouteStatus,
+			VolumesStatus:       capp.Status.VolumesStatus,
+			Conditions:          capp.Status.Conditions,
+		},
+	}
+}
+
+func convertV1alpha1CappToTypesCapp(v1Capp v1alpha1.Capp) types.Capp {
+	return types.Capp{
+		Metadata: types.Metadata{
+			Name:      v1Capp.Name,
+			Namespace: v1Capp.Namespace,
+		},
+		Annotations: convertMapToAnnotations(v1Capp.Annotations),
+		Labels:      convertMapToLabels(v1Capp.Labels),
+		Spec: v1alpha1.CappSpec{
+			ScaleMetric:       v1Capp.Spec.ScaleMetric,
+			Site:              v1Capp.Spec.Site,
+			State:             v1Capp.Spec.State,
+			ConfigurationSpec: v1Capp.Spec.ConfigurationSpec,
+			RouteSpec:         v1Capp.Spec.RouteSpec,
+			LogSpec:           v1Capp.Spec.LogSpec,
+			VolumesSpec:       v1Capp.Spec.VolumesSpec,
+		},
+		Status: v1alpha1.CappStatus{
+			ApplicationLinks:    v1Capp.Status.ApplicationLinks,
+			KnativeObjectStatus: v1Capp.Status.KnativeObjectStatus,
+			RevisionInfo:        v1Capp.Status.RevisionInfo,
+			StateStatus:         v1Capp.Status.StateStatus,
+			LoggingStatus:       v1Capp.Status.LoggingStatus,
+			RouteStatus:         v1Capp.Status.RouteStatus,
+			VolumesStatus:       v1Capp.Status.VolumesStatus,
+			Conditions:          v1Capp.Status.Conditions,
+		},
+	}
+}
+
+func convertAnnotationsToMap(kvList []types.KeyValue) map[string]string {
+	annotations := make(map[string]string)
+	for _, kv := range kvList {
+		annotations[kv.Key] = kv.Value
+	}
+	return annotations
+}
+
+func convertMapToAnnotations(annotations map[string]string) []types.KeyValue {
+	var kvList []types.KeyValue
+	for key, value := range annotations {
+		kvList = append(kvList, types.KeyValue{Key: key, Value: value})
+	}
+	return kvList
+}
+
+func convertMapToLabels(labels map[string]string) []types.KeyValue {
+	var kvList []types.KeyValue
+	for key, value := range labels {
+		kvList = append(kvList, types.KeyValue{Key: key, Value: value})
+	}
+	return kvList
+}
+
+func convertLabelsToMap(kvList []types.KeyValue) map[string]string {
+	labels := make(map[string]string)
+	for _, kv := range kvList {
+		labels[kv.Key] = kv.Value
+	}
+	return labels
+}
```

Comment:
I think all of those will prove useful in other resources, so maybe move them to a different file?

Reply:
I disagree, it's not used in this PR outside `containerapp.go` so let's keep it only here.